src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceMarginViewMargin.cs

```diff
@@ -0,0 +1,238 @@
+            var endOfChangedSpan = changedSnapshotSpans.Max(span => span.End);
+            var changedSpan = new SnapshotSpan(startOfChangedSpan, endOfChangedSpan);
+
+            _glyphManager.RemoveGlyph(changedSpan);
+
+            foreach (var line in _textView.TextViewLines.GetTextViewLinesIntersectingSpan(changedSpan))
+            {
+                if (line.IsValid)
+                {
+                    RefreshGlyphsOver(line);
+                }
+            }
+        }
+
+        private void RefreshGlyphsOver(ITextViewLine textViewLine)
+        {
+            if (!_optionService.GetOption(FeatureOnOffOptions.InheritanceMarginCombinedWithIndicatorMargin, _languageName))
+            {
+                foreach (var mappingTagSpan in _tagAggregator.GetTags(textViewLine.ExtentAsMappingSpan))
+                {
+                    // Only take tag spans with a visible start point and that map to something
+                    // in the edit buffer and *start* on this line
+                    if (mappingTagSpan.Span.Start.GetPoint(_textView.VisualSnapshot.TextBuffer, PositionAffinity.Predecessor) != null)
+                    {
+                        var tagSpans = mappingTagSpan.Span.GetSpans(_textView.TextSnapshot);
+                        if (tagSpans.Count > 0)
+                        {
+                            _glyphManager.AddGlyph(mappingTagSpan.Tag, tagSpans[0]);
+                        }
+                    }
+                }
+            }
+        }
+
+        #region IWpfTextViewMargin
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(InheritanceMarginViewMargin));
+            }
+        }
+
+        public FrameworkElement VisualElement
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _grid;
+            }
+        }
+
+        public double MarginSize
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _grid.ActualWidth;
+            }
+        }
+
+        public bool Enabled
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return true;
+            }
+        }
+
+        public ITextViewMargin? GetTextViewMargin(string marginName) =>
+            marginName == nameof(InheritanceMarginViewMargin) ? this : null;
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _tagAggregator.BatchedTagsChanged -= OnTagsChanged;
+                _textView.LayoutChanged -= OnLayoutChanged;
+                _textView.ZoomLevelChanged -= OnZoomLevelChanged;
+                _optionService.OptionChanged -= OnRoslynOptionChanged;
+                _tagAggregator.Dispose();
+                _glyphManager.Dispose();
+                _disposed = true;
```

Comment:
i recommend doing this first.  that way none of the operations after can cause reentrancy.